### PR TITLE
Add Temporal-Item-Frequency-based User-KNN (TIFUKNN) model

### PR DESCRIPTION
**Pull request by lthoang**
### Checklist:

- [ ] I have added tests.
- [ ] I have updated the documentation accordingly.
- [x] I have updated `README.md` (if you are adding a new model).
- [x] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

File: cornac/models/tifuknn/recom_tifuknn.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+    def fit(self, train_set, val_set=None):
+        from sklearn.neighbors import NearestNeighbors
+
+        super().fit(train_set=train_set, val_set=val_set)
+        self.user_vectors = self._get_user_vectors(self.train_set)
+        if self.n_neighbors > len(self.user_vectors):
+            warnings.warn("Number of users is %d, smaller than number of neighbors %d" % (len(self.user_vectors), self.n_neighbors))
+            self.n_neighbors = len(self.user_vectors)
+        self.nbrs = NearestNeighbors(n_neighbors=self.n_neighbors, algorithm="brute").fit(list(self.user_vectors.values()))
```
Comment (tqtg) [inline cornac/models/tifuknn/recom_tifuknn.py:85]: If it's really brute force search, do we even need sklearn for this? We can just use numpy to compute l2 distances and find nearest neighbors in the score function.

Comment (tqtg) [reply cornac/models/tifuknn/recom_tifuknn.py:85]: Alternatively, [KDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.query.html) in scipy is generally faster than the KNN implementation in sklearn. Another advantage is that we already have scipy as dependency. Just something to consider.

File: cornac/models/tifuknn/recom_tifuknn.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        return self
+
+    def _get_user_vectors(self, data_set):
+        user_vectors = {}
+        for [user_idx], _, [basket_items] in tqdm(
+            data_set.ubi_iter(batch_size=1, shuffle=False),
+            desc="Getting user vectors",
+            total=data_set.num_users,
+        ):
+            user_vectors[user_idx] = self._compute_user_vector(basket_items[:-1])
```
Comment (tqtg) [inline cornac/models/tifuknn/recom_tifuknn.py:103]: do we need `user_vectors` to be a dic? Is it possible for it to be a numpy 2D array?

File: cornac/models/tifuknn/recom_tifuknn.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                        grouped_vec_list.append(sum / est_num)
+
+            return grouped_vec_list, n_groups
+
+    def score(self, user_idx, history_baskets, **kwargs):
+        if len(history_baskets) == 0:
+            return np.zeros(self.total_items, dtype="float32")
+        user_vector = self._compute_user_vector(history_baskets)
+        _, indices = self.tree.query([user_vector], k=self.n_neighbors)
+        return self.alpha * user_vector + (1 - self.alpha) * np.sum([self.user_vectors[idx] for idx in indices.squeeze()], axis=0)
```
Comment (tqtg) [inline cornac/models/tifuknn/recom_tifuknn.py:171]: Based on Section 3.1 in the paper, $\textbf{u}_n$ should be the average vector of neighbors, not sum. Please double check on that. Also, we simply pass the indices directly instead of doing for loop:
```python
np.mean(self.user_vectors[indices.squeeze()])
```

Comment (tqtg) [reply cornac/models/tifuknn/recom_tifuknn.py:171]: @lthoang

Comment (lthoang) [reply cornac/models/tifuknn/recom_tifuknn.py:171]: I've  double checked. It should be the average vector of neighbors. I will make update in the next commit.

**Review by tqtg — APPROVED** [review]

**Comment by tqtg** [comment]
LGTM
